### DISCUSSION TITLE
Alternatives to  Modeling AWS Karpenter Resources without writing Custom Deserializer/s

### DISCUSSION THREAD

**Discussion by abrarw007** [fabric8io/kubernetes-client#5964] [General]
Hello, Fabric8 community!

I'm working on a project where I need to fetch the AWS Karpenter resources from my Kubernetes clusters using the Fabric8 GenericKubernetesResourceList model. While I can successfully extract the required JSON spec and other attributes from this model, I'm facing a challenge when it comes to modeling this data into a custom Karpenter class.

Currently, I'm relying on writing a custom Deserializer for my custom Karpenter class to create Karpenter objects from the extracted data. However, I'd like to explore if there's a built-in way within Fabric8 to achieve this without having to implement a custom Deserializer.

Could someone please guide me on whether there's an alternative approach or a built-in feature in Fabric8 that could help me create instances of my custom Karpenter class directly from the GenericKubernetesResourceList model, without the need for a custom Deserializer?

Any insights or suggestions would be greatly appreciated. Thank you in advance for your help!

**Comment by rohanKanojia** [comment]
@abrarw007 : Are you trying to generate Karpenter CustomResource POJOs from CRD YAML? If yes, have you checked [Fabric8 Java Generator](https://github.com/fabric8io/kubernetes-client/blob/main/doc/java-generation-from-CRD.md)?

**Reply by abrarw007** [reply]
Hey @rohanKanojia , thank you for your response. I hadn't checked this. 
I don't think that approach works for my use case because I can't create a YAML file for every Karpenter resource. Additionally, if my application is running in someone else's Kubernetes cluster, I would have to go over every Karpenter Autoscaler and create YAML files from the data returned by the Kubernetes client. 
That seems like too many added steps if I'm trying to avoid writing a custom Deserializer.
I was hoping to find a more straightforward way within Fabric8 to instantiate my custom Karpenter class directly from the GenericKubernetesResourceList model, which represents the  Karpenter Autoscalar deployed in the cluster, without the need for a custom Deserializer or having to generate yaml files.

**Reply by rohanKanojia** [reply]
@abrarw007 : I'm not sure I'm understanding your use case correctly. I was referring to CustomResourceDefinition YAML, not CustomResource YAML. I think you should be able to generate Karpenter model objects by providing these CRD as input to java generator:

- https://github.com/aws/karpenter-provider-aws/blob/5f5bba496bc7debb4c235382a3a7d03fee30da44/pkg/apis/crds/karpenter.sh_nodepools.yaml#L4
- https://github.com/aws/karpenter-provider-aws/blob/5f5bba496bc7debb4c235382a3a7d03fee30da44/pkg/apis/crds/karpenter.k8s.aws_ec2nodeclasses.yaml
- https://github.com/aws/karpenter-provider-aws/blob/5f5bba496bc7debb4c235382a3a7d03fee30da44/pkg/apis/crds/karpenter.sh_nodeclaims.yaml


Please see this [blogpost](https://developers.redhat.com/articles/2023/01/24/how-generate-code-using-fabric8-kubernetes-client) for more details.

In order to use typed api, you either need to write the Karpenter model classes yourself or generate it using Fabric8 Java Generator.
